posts/2021/03/array_interchange_protocols.md

```diff
@@ -0,0 +1,370 @@
+<!--
+.. title: Guide to array, tensor and buffer protocols in Python
+.. slug: array-interchange-protocols
+.. date: 2021-03-07 08:00:00 UTC-00:00
+.. author: Pearu Peterson, Ralf Gommers
+.. tags: Labs, NumPy, Numba, PyTorch, data-apis
+.. category:
+.. link:
+.. description:
+.. type: text
+-->
+
+The aim of this blog post is to review the current state of CPU/CUDA Array Interfaces and PEP 3118 Buffer Protocol
+in the context of NumPy and PyTorch, and give recommendations to improve the PyTorch support to these protocols.
+We also indicate the overall usage of array interfaces in different Python libraries.
+
+This blog post is inspired by a [PyTorch issue 51156](https://github.com/pytorch/pytorch/issues/51156).
+
+TODO: rewrite the above a bit, and mention these are all zero-copy, in-process protocols. For serialization,
+see elsewhere (not sure where).
+
+
+<!-- TEASER_END -->
+
+## Summary of protocols
+
+### CPU Array Interface
+
+[Array Interface (Version
+3)](https://numpy.org/doc/stable/reference/arrays.interface.html)
+defines a protocol for objects to re-use each other's data buffers.
+It was created in 2005 within the [NumPy](https://numpy.org/) project for CPU array-like
+objects. The implementation of the array interface is defined by the
+existence of the following attributes:
+
+- `__array_interface__` - a Python dictionary that contains the shape,
+  the element type, and optionally, the data buffer address and the
+  strides of an array-like object.
+
+- `__array_struct__` - holds a pointer to [PyArrayInterface
+  C-structure](https://numpy.org/doc/stable/reference/arrays.interface.html#object.__array_struct__)
+  which can mediate any data buffer implementation (say, written in
+  C/C++) to Python via NumPy ndarray object.
```

Comment:
I'm actually not 100% sure: is there any reason today to use this over the buffer protocol?

Reply:
Perhaps not, but the `__array_struct__` exists today.

Comparing https://www.python.org/dev/peps/pep-3118/#the-py-buffer-struct and https://numpy.org/doc/stable/reference/arrays.interface.html#object.__array_struct__, indeed, the buffer struct is less tied to Python C/API and therefore we should recommend using it (the buffer protocol).